onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQAttention.cpp

```diff
@@ -0,0 +1,656 @@
+
+        DML_ELEMENT_WISE_DEQUANTIZE_LINEAR_OPERATOR_DESC inputDequantizeOperatorDesc = {};
+        inputDequantizeOperatorDesc.InputTensor = &inputDescs[InputIndex::inputIndex];
+        inputDequantizeOperatorDesc.ScaleTensor = &inputDescs[InputIndex::inputScaleIndex];
+        inputDequantizeOperatorDesc.ZeroPointTensor = &inputDescs[InputIndex::inputZeroPointIndex];
+        inputDequantizeOperatorDesc.OutputTensor = &namedIntermediateInputTensorDesc;
+
+        const DML_OPERATOR_DESC inputDequantizeOpDesc{DML_OPERATOR_ELEMENT_WISE_DEQUANTIZE_LINEAR, &inputDequantizeOperatorDesc};
+
+        DML_ELEMENT_WISE_DEQUANTIZE_LINEAR_OPERATOR_DESC weightDequantizeOperatorDesc = {};
+        weightDequantizeOperatorDesc.InputTensor = &inputDescs[InputIndex::weightsIndex];
+        weightDequantizeOperatorDesc.ScaleTensor = &inputDescs[InputIndex::weightScaleIndex];
+        weightDequantizeOperatorDesc.ZeroPointTensor = &inputDescs[InputIndex::weightZeroPointIndex];
+        weightDequantizeOperatorDesc.OutputTensor = &namedIntermediateWeightTensorDesc;
+
+        const DML_OPERATOR_DESC weightDequantizeOpDesc{DML_OPERATOR_ELEMENT_WISE_DEQUANTIZE_LINEAR, &weightDequantizeOperatorDesc};
+
+        DML_GEMM_OPERATOR_DESC gemmOperatorDesc = {};
+        gemmOperatorDesc.ATensor = inputDequantizeOperatorDesc.OutputTensor;
+        gemmOperatorDesc.BTensor = weightDequantizeOperatorDesc.OutputTensor;
+
+        if (hasBias)
+        {
+            gemmOperatorDesc.CTensor = &inputDescs[2];
+        }
+
+        gemmOperatorDesc.OutputTensor = &namedFirstGemmOutputTensorDesc;
+        gemmOperatorDesc.TransA = DML_MATRIX_TRANSFORM_NONE;
+        gemmOperatorDesc.TransB = DML_MATRIX_TRANSFORM_NONE;
+        gemmOperatorDesc.Alpha = 1.0f;
+        gemmOperatorDesc.Beta = 1.0f;
+        gemmOperatorDesc.FusedActivation = nullptr;
+        const DML_OPERATOR_DESC gemmDesc {DML_OPERATOR_GEMM, &gemmOperatorDesc};
+
+        std::array<uint32_t, 3> queryKeySlicedTensorShape {batchSize, sequenceLength, hiddenSize + hiddenSize};
+        TensorDesc queryKeySlicedInputTensorDesc = TensorDesc::ConstructDefaultTensorDesc(dataType, queryKeySlicedTensorShape);
+        DML_TENSOR_DESC namedQueryKeySlicedInputTensorDesc = queryKeySlicedInputTensorDesc.GetDmlDesc();
+
+        std::array<uint32_t, 3> valueSlicedTensorShape {batchSize, sequenceLength, vHiddenSize};
+        TensorDesc valueSlicedInputTensorDesc = TensorDesc::ConstructDefaultTensorDesc(dataType, valueSlicedTensorShape);
+        DML_TENSOR_DESC namedValueSlicedInputTensorDesc = valueSlicedInputTensorDesc.GetDmlDesc();
+
+        // Transpose slice QK from [batchSize, sequenceLength, 2, numHeads, headSize] to [batchSize, sequenceLength, numHeads, 2, headSize]
+        std::array<uint32_t, 5> queryKeyTransposedTensorShape {batchSize, sequenceLength, numHeads, 2, headSize};
+        std::array<uint32_t, 5> queryKeyTransposedStrides {
```

Comment:
```suggestion
        std::array<uint32_t, 5> queryKeyTransposedStrides = {
```

[nit] Using a mix of "=" and no "=" for initialization is odd. e.g. Above we see:

`DML_TENSOR_DESC namedQueryKeySlicedInputTensorDesc = queryKeySlicedInputTensorDesc.GetDmlDesc();`

Rather than:

`DML_TENSOR_DESC namedQueryKeySlicedInputTensorDesc { queryKeySlicedInputTensorDesc.GetDmlDesc() };`